### PR TITLE
Solution

### PR DESCRIPTION
**Pull request by Hradivnyk**
[DEMO LINK](https://Sholudyvyy.github.io/react_todo-app-with-api/)

### CODE REVIEW

File: src/api/todos.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+export const deleteTodo = (todoId: number) => {
+  return client.delete(`/todos/${todoId}`);
+};
+
+export const updateTodoTitle = (todoId: number, newTitle: string) => {
+  return client.patch(`/todos/${todoId}`, { title: newTitle });
+};
+
+export const updateTodoCompleted = (todoId: number, isCompleted: boolean) => {
+  return client.patch(`/todos/${todoId}`, { completed: isCompleted });
```
Comment (RostyslavSharuiev) [inline src/api/todos.ts:28]: Looks good, but you can combine `updateTodoTitle` and `updateTodoCompleted` in one function

File: src/components/TodoHeader/TodoHeader.tsx [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+  UpdateAllTodosCompleted,
+}) => {
+  const [inputText, setInputText] = useState('');
+  const isAllTodosCompleted = countOfCompletedTodos === countOfTodos;
+  const inputRef = useRef<HTMLInputElement>(null);
+
+  async function handleSubmit(event: React.FormEvent<HTMLFormElement>) {
+    event.preventDefault();
+
+    if (inputText.trim()) {
```
Comment (RostyslavSharuiev) [inline src/components/TodoHeader/TodoHeader.tsx]: For `inputText.trim()` create variable `formattedTitle` and use it for `if` statements

File: src/components/TodoTask/TodoTask.tsx [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    if (event.key === 'Escape') {
+      setTempTodoTitle(todo.title);
+      setIsEditing(false);
+    }
+  };
+
+  async function updateTitle() {
+    const trimNewTitle = tempTodoTitle.trim();
+
+    if (!trimNewTitle.length) {
```
Comment (RostyslavSharuiev) [inline src/components/TodoTask/TodoTask.tsx]: `!'' === !''.length`

File: src/components/TodoTask/TodoTask.tsx [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+      onDeleteTodo(todo.id);
+    }
+
+    if (todo.title !== trimNewTitle) {
+      if (await fetchUpdateTodoTitle(todo.id, tempTodoTitle)) {
```
Comment (RostyslavSharuiev) [inline src/components/TodoTask/TodoTask.tsx]: `const {id} = todo`, so try destructurise it

**Review by RostyslavSharuiev — APPROVED** [review]

File: src/App.tsx [hunk 5]
```diff
@@ -1,26 +1,205 @@
...
+        }),
+      );
+    } catch {
+      setErrorMessage(Errors.updateError);
+    } finally {
+      setUpdatedTodosId(current => current.filter(id => id !== todoId));
+    }
+  }
+
+  async function UpdateAllTodosCompleted() {
```
Comment (igorkruz) [inline src/App.tsx]: Rename `UpdateAllTodosCompleted `to `updateAllTodosCompletion`
This makes it clear that the function is intended to update the completion status of all todos.

File: src/App.tsx [hunk 6]
```diff
@@ -1,26 +1,205 @@
...
+      <h1 className="todoapp__title">todos</h1>
+
+      <div className="todoapp__content">
+        <TodoHeader
+          countOfTodos={todos.length}
+          countOfCompletedTodos={countOfCompletedTodos}
+          isInputDisabled={!!tempTodo}
+          fetchAddTodo={fetchAddTodo}
+          setErrorMessage={setErrorMessage}
+          UpdateAllTodosCompleted={UpdateAllTodosCompleted}
```
Comment (igorkruz) [inline src/App.tsx]: When passing props from a parent component to a child component, it's important to follow clear naming conventions that convey the purpose of the props, especially for handler functions.

Use handle prefix for event handlers in the parent and on prefix in the child.

File: src/components/TodoHeader/TodoHeader.tsx [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+  async function handleSubmit(event: React.FormEvent<HTMLFormElement>) {
+    event.preventDefault();
+
+    const formattedTitle = inputText.trim();
+
+    if (formattedTitle) {
+      if (await fetchAddTodo(formattedTitle)) {
+        setInputText('');
+      }
```
Comment (igorkruz) [inline src/components/TodoHeader/TodoHeader.tsx]: The function is well-structured, but consider adding a try-catch block for error handling during the fetch call.

**Review by igorkruz — CHANGES_REQUESTED** [review]
Good Job 👍 
let's improve your solution

File: src/App.tsx [hunk 8]
```diff
@@ -1,26 +1,207 @@
...
+import { Errors } from './types/Errors';
 
 export const App: React.FC = () => {
-  if (!USER_ID) {
-    return <UserWarning />;
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [completedCategory, setCompletedCategory] =
+    useState<TodoCompletedCategory>(TodoCompletedCategory.all);
+  const [errorMessage, setErrorMessage] = useState<string>(Errors.noneError);
```
Comment (VolodymyrKirichenko) [inline src/App.tsx]: use `enum` as Type
```suggestion
  const [errorMessage, setErrorMessage] = useState<Errors>(Errors.noneError);
```

File: src/App.tsx [hunk 9]
```diff
@@ -1,26 +1,207 @@
...
+  const [updatedTodosId, setUpdatedTodosId] = useState([0]);
+
+  const filtredTodos = filterTodosByComplated(todos, completedCategory);
```
Comment (VolodymyrKirichenko) [inline src/App.tsx]: use `useMemo` for all such cases

File: src/App.tsx [hunk 10]
```diff
@@ -1,26 +1,207 @@
...
+    try {
+      const todosFromApi = await getTodos();
+
+      setTodos(todosFromApi);
+    } catch {
+      setErrorMessage(Errors.loadError);
+    }
+  }
+
+  async function handleAddTodo(newTitle: string) {
```
Comment (VolodymyrKirichenko) [inline src/App.tsx]: use `useCallback` for all such cases(when u pass it to another component like props)

File: src/api/todos.ts [hunk 11]
```diff
@@ -0,0 +1,30 @@
+import { Todo } from '../types/Todo';
+import { client } from '../utils/fetchClient';
+
+export const USER_ID = 1617;
```
Comment (VolodymyrKirichenko) [inline src/api/todos.ts]: move CONSTANTS to separate - `utils/constants.ts` file

File: src/api/todos.ts (hunk 1 above)
Comment (VolodymyrKirichenko) [reply src/api/todos.ts:28]: Good point, but up to u)

File: src/components/ErrorNotification/ErrorNotification.tsx [hunk 12]
```diff
@@ -0,0 +1,43 @@
+import React, { useEffect } from 'react';
+import { Errors } from '../../types/Errors';
+import classNames from 'classnames';
+
+type Props = {
+  errorMessage: string;
```
Comment (VolodymyrKirichenko) [inline src/components/ErrorNotification/ErrorNotification.tsx:6]: use ur `enum` Errors
```suggestion
  errorMessage: Errors;
```

File: src/components/TodoFooter/TodoFooter.tsx [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+
+type Props = {
+  countOfNotCompletedTodos: number;
+  isSomeTodoComplated: boolean;
+  completedCategory: CompletedCategory;
+  onCompletedCategory: (completedCategory: CompletedCategory) => void;
+  deleteAllCompletedTodos: () => void;
+};
+
+const activityFilters = {
```
Comment (VolodymyrKirichenko) [inline src/components/TodoFooter/TodoFooter.tsx]: move it to a separate file - `utils/consts.ts`

File: src/components/TodoFooter/TodoFooter.tsx [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+          </a>
+        ))}
+      </nav>
+
+      <button
+        type="button"
+        className="todoapp__clear-completed"
+        data-cy="ClearCompletedButton"
+        disabled={!isSomeTodoComplated}
+        onClick={() => deleteAllCompletedTodos()}
```
Comment (VolodymyrKirichenko) [inline src/components/TodoFooter/TodoFooter.tsx]: why not?
```suggestion
        onClick={deleteAllCompletedTodos}
```

**Review by VolodymyrKirichenko — CHANGES_REQUESTED** [review]
Good job, but let's make some changes 🥹

File: src/App.tsx (hunk 10 above)
Comment (Hradivnyk) [reply src/App.tsx]: when I don't pass it to another component I don't use `useCallback`?

Comment (KosolapovAnatolii) [reply src/App.tsx]: When you don't pass func as props to child component it's up to you.
But when you pass, it's prevent extra rerenders of you child component, that's why it's important

File: src/components/ErrorNotification/ErrorNotification.tsx (hunk 12 above)
Comment (KosolapovAnatolii) [reply src/components/ErrorNotification/ErrorNotification.tsx:6]: Didn't fixed

**Review by KosolapovAnatolii — APPROVED** [review]
LGTM 🆒 
Don't forget fix small issues that I noticed and go forward to the next tasks 💪

**Pull request closed without merging**